### PR TITLE
router:add retry stat for completed requests

### LINKED ISSUE #7670: some retries are skipped by Envoy
I am debugging some retry failures and noticed that Envoy is skipping retry in some cases even though the retry policy is set for it. This is a gRPC retry.

Here is the retry configuration
num_retries: 5
retry-on: cancelled,connect-failure,gateway-error,refused-stream,resource-exhausted,unavailable

Actual end client sees 503 with error message "upstream connect error or disconnect/reset before headers. reset reason: connection failure"

Here is some relevant stats : 
envoy.cluster.upstream_rq_retry : 7
envoy.cluster.upstream_rq_retry_success: 6
envoy.cluster.upstream_rq_503: 1
envoy.http2.tx_reset:1
envoy.cluster.upstream_rq_pending_failure_eject:7
envoy.cluster.upstream_cx_connect_fail:7
envoy.cluster.upstream_cx_connect_timeout:3

As can be seen from the above stats, one retry is skipped and resulted in 503. 

Looking at the `router` code, I see two places where a retry can be skipped on upstream reset.
One here, But since it is a connection failure - none these conditions should trigger the skip. Also `rq_reset_after_downstream_response_started` stat is not incremented. So I am guessing this is not the case. 
https://github.com/envoyproxy/envoy/blob/7e7486588239f42b4be3a6383e3c0776d2d05bef/source/common/router/router.cc#L850

Second here : 
https://github.com/envoyproxy/envoy/blob/7e7486588239f42b4be3a6383e3c0776d2d05bef/source/common/router/router.cc#L1218

I am not able to confirm whether this is case, because there are no stats to confirm the same. But the comments above seems very relevant on why retry may have been skipped.

One thing we can try is increase connect timeout to see if it resolves. Another thing I want to check, can adding buffer filter before this help?

### PR DESCRIPTION
**Pull request by ramaraochavali**
Description: Adds a router stat that indicates the retry is skipped because of completed request.
Risk Level: Low
Testing: Added
Docs Changes: Updated
Release Notes: Added
Fixes #7670

### CODE REVIEW

File: source/common/router/router.cc [hunk 1]
```diff
@@ -1216,6 +1216,7 @@ bool Filter::setupRetry() {
   // this filter which will make this a non-issue. The implementation of supporting retry in cases
   // where the request is not complete is more complicated so we will start with this for now.
   if (!downstream_end_stream_) {
+    config_.stats_.rq_retry_skipped_after_request_completed_.inc();
```
Comment (snowp) [inline source/common/router/router.cc]: Isn't this checking that we've fully received the request? Would it make more sense word this as `rq_retry_skipped_request_not_complete` or something to that effect?

Comment (ramaraochavali) [reply source/common/router/router.cc]: Yeah. That makes sense

**Comment by ramaraochavali** [comment]
@snowp changed the stat name and docs. PTAL

**Pull request merged**